File: ruby/easy/valid_palindrome.rb
# @param {String} s
# @return {Boolean}
def is_palindrome(s)
    i,j = 0,s.length-1
    while i <= j
        i += 1 and next unless is_char(s[i])
        j -= 1 and next unless is_char(s[j])

        return false unless char_equals(s[i], s[j])
        i += 1 and j -= 1
    end
    return true
end

def is_char(c)
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')
end

def char_equals(c1, c2)
    return c1.upcase == c2.upcase
end

File: javascript/easy/validPalindrome.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
    let i = 0, j = s.length-1;
    while(i <= j) {
        if(!isChar(s[i])) {
            i++; continue;
        }

        if(!isChar(s[j])) {
            j--; continue;
        }

        if(!charEquals(s[i], s[j])) {
            return false;
        }
        i++; j--;
    }
    return true;
};

var isChar = function(c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9');
};

var charEquals = function(c1, c2) {
    return c1.toUpperCase() === c2.toUpperCase();
};
